


# PRS-Bridge-article-code

This repository contains instruction and scripts to reproduce all the synthetic and real data analysis results in the PRS-Bridge article, benchmarking the following methods: PRS-Bridge, LDpred2, LASSOSUM, and PRS-CS, PRS-CS-Projection, PRS-CS-threshold, and PRS-CS-Regularized.


## Prerequisites
-   **External software:**  PRS-Bridge [https://github.com/YuzhengDun1999/PRSBridge](https://github.com/YuzhengDun1999/PRSBridge); PRS-CS [https://github.com/getian107/PRScs](https://github.com/getian107/PRScs); Plink 1.9 [https://www.cog-genomics.org/plink/](https://www.cog-genomics.org/plink/) and 2.0 [https://www.cog-genomics.org/plink/2.0/](https://www.cog-genomics.org/plink/2.0/); LDpred2 [https://privefl.github.io/bigsnpr/articles/LDpred2.html](https://privefl.github.io/bigsnpr/articles/LDpred2.html); LASSOSUM [https://github.com/tshmak/lassosum](https://github.com/tshmak/lassosum).
-   **Data:**  UK Biobank access for real trait analyses; simulated data and summary statistics for disease traits public available.

## Data Download
-   **Simulated data:**  
    Download from  [https://doi.org/10.7910/DVN/COXHAP](https://doi.org/10.7910/DVN/COXHAP).
-   **Real data:**
    -   Disease traits: check our manuscript and acc_form for reference and download link.
    -   Continuous traits: Individual-level genotype and phenotype data UK Biobank (application required). Follow the instruction in run-real-data-analysis/process_continuous_dat.R to get summary statistics.
 
## Directory Contents and Usage

### data
This folder contains the data used to generate posterior samples of the regression coefficients used to create Figure 1, as well as an example summary-statistics file provided for illustration only. Specifically, the files included are:
-   `PRScs_sumdat.txt`,  `chr22.bim`: For Figure 1.
-   `sumdat_Rcov.txt`: Example input summary statistics used by the preprocessing scripts in the run-real-data-analysis/ directory to demonstrate how input data should be formatted before preparing data for running all methods.

### PRS-CS-proj

This folder provides our modification of the PRS-CS code, the original version of which is found at [https://github.com/getian107/PRScs](https://github.com/getian107/PRScs). 
The files included are:

-   `PRScs_noconverge.py`: Implements a modified version of PRS-CS that removes the ad hoc constraint on the prior variance. 
See `plot-results/plot_fig1.R` for example usage.
- `PRScs_threshold.py`: Implements versions of PRS-CS modified to use different values of the upper threshold on the prior variance. 
See the scripts files generated by `run-real-data-analysis/PRScs_threshold.R` for example usage.
-   `PRScs_proj.py`: Implements versions of PRS-CS modified to use projected summary statistics. 
See the scripts files generated by `run-real-data-analysis/PRScs_proj.R` for example usage.
-   `PRScs_Regularized.py`: Implements versions of PRS-CS modified to use Regularized LD matrix. 
See the scripts files generated by `run-real-data-analysis/PRScs_Regularized.R` for example usage.

### simulation

This folder includes scripts to reproduce the numerical results of Section 4.1 based on the synthetic data.
The steps to reproduce the results are as follows:

1.  **Download simulated data**  (see above)
2.  **Preprocess summary statistics:** Run `process_simulated_dat.R`; detailed explanations of the input parameters are provided within the script.
3.  **Run PRS methods:**  Run `PRSBridge.R`, `PRScs.R`, and `ldpred2.R`; the input parameters are the same as those in `process_simulated_dat.R`. Please see each method's tutorial on how to estimate or download LD reference panel.
4.  **Evaluate results:** Run `evaluation.R`; input parameters are the same as `process_simulated_dat.R`.
5.  **Visualize:**  Run `plot.R` to produce Figure 3.

### run-real-data-analysis

These scripts are used to run each method on the real-data summary statistics. Note that for PRS-Bridge and PRS-CS, it is not feasible to reproduce the full results on a local machine due to computational demands. Instead, running the R scripts will generate a set of .sh files that can be submitted to a high-performance computing cluster (HPC) using SLURM. The command line for running each method are specified within the generated .sh files. For LASSOSUM and ldpred2, we also provide scripts for running the methods; however, we recommend executing them on an HPC system as well.

**Step-by-step workflow:**

1.  **Preprocess GWAS summary statistics:**
    -   Disease traits: run  `process_disease_dat.R`
    -   Continuous traits (UK Biobank): run  `process_continuous_dat.R` to get summary statistics
2.  **Quality control:**  run  `sumdat_QC.R`
3.  **Run each PRS method:**  
    Use method-specific scripts (e.g.,  `PRScs.R`,  `PRSBridge.R`, etc.), ensuring the correct LD reference panel is specified (see each method's tutorial on how to estimate or download LD reference panel).

### evaluation-real-data-analysis
Scripts for evaluating methods:

-   `RE_Lassosum.R`: Computes relative efficiency for each method.
-   `get_sd_*.R`: Computes $R^2$ (or transformed AUC) and standard errors for each method:
    -   `get_sd_PRSBridge.R`
    -   `get_sd_PRScs.R`
    - ...

### plot-results

Scripts for visualization of results and reproducing figures:

-   `plot.R`: Generates results for Figures 4 and 5.
-   `plot_fig1.R`: Reproduces Figure 1 (posterior coefficients).
-   All plot scripts are annotated to specify which figure they reproduce and what input parameters to use in scripts in run-real-data-analysis section.
